### PR TITLE
added a new feature to scan folders where entities are located

### PR DESCRIPTION
**Pull request by flug**
| Q             | A
| ------------- | ---
| Bug fix?      | no 
| New feature?  | yes 
| BC breaks?    | no
| Deprecations? | no
| Tickets       | fixes #... 
| License       | MIT
| Doc PR        | api-platform/docs#...

### CODE REVIEW

**Comment by teohhanhui** [comment]
What is this for? We already support loading from directories, be it XML, YAML or annotations.

**Comment by flug** [comment]
@soyuka wanted me to bring this feature to see with him and @dunglas

**Comment by dunglas** [comment]
@teohhanhui Basically with this and `defaults`, you don't need the any annotations to expose a simple API, just pure PHP classes.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -177,8 +177,9 @@ public function getConfigTreeBuilder()
                     ->addDefaultsIfNotSet()
                     ->children()
                         ->arrayNode('paths')
-                            ->prototype('scalar')->end()
-                        ->end()
+                        ->prototype('scalar')->end()
+                    ->end()
```
Comment (alanpoulain) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Bad indentation.

File: src/Bridge/Symfony/Bundle/Resources/config/metadata/directory.xml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services">
+    <services>
+        <service id="api_platform.metadata.resource.name_collection_factory.directory" decorates="api_platform.metadata.resource.name_collection_factory"
+                 class="ApiPlatform\Core\Metadata\Resource\Factory\DirectoryResourceNameCollectionFactory" public="false">
+           <argument>%api_platform.resource_class_directories%</argument>
```
Comment (alanpoulain) [inline src/Bridge/Symfony/Bundle/Resources/config/metadata/directory.xml]:
```suggestion
            <argument>%api_platform.resource_class_directories%</argument>
```

File: src/Metadata/Extractor/ClassExtractor.php [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Metadata\Extractor;
+
+use ApiPlatform\Core\Util\ReflectionClassRecursiveIterator;
+use Doctrine\Common\Annotations\Reader;
+use Doctrine\ORM\Mapping as ORM;
+use Psr\Container\ContainerInterface;
+
+class ClassExtractor extends AbstractExtractor
```
Comment (alanpoulain) [inline src/Metadata/Extractor/ClassExtractor.php]: final?

File: src/Metadata/Extractor/ClassExtractor.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    public function __construct(array $paths, ContainerInterface $container = null, Reader $reader)
+    {
+        $this->reader = $reader;
+        parent::__construct($paths, $container);
+    }
+
+    protected function extractPath(string $path)
+    {
+        try {
```
Comment (alanpoulain) [inline src/Metadata/Extractor/ClassExtractor.php]: ?

File: src/Metadata/Extractor/ClassExtractor.php [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+{
+    private $reader;
+
+    public function __construct(array $paths, ContainerInterface $container = null, Reader $reader)
+    {
+        $this->reader = $reader;
+        parent::__construct($paths, $container);
+    }
+
+    protected function extractPath(string $path)
```
Comment (alanpoulain) [inline src/Metadata/Extractor/ClassExtractor.php]:
```suggestion
    protected function extractPath(string $path): void
```

File: src/Metadata/Extractor/ClassExtractor.php [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        try {
+        } catch (\InvalidArgumentException $e) {
+            throw new \InvalidArgumentException($e->getMessage(), $e->getCode(), $e);
+        }
+        if ('' === $path) {
+            return;
+        }
+        /**
+         * @var string
```
Comment (alanpoulain) [inline src/Metadata/Extractor/ClassExtractor.php]: To remove.

File: src/Metadata/Extractor/ClassExtractor.php [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                'attributes' => null,
+                'identifier' => $this->isIdentifier($property),
+                'subresource' => null,
+            ];
+        }
+
+        return $properties;
+    }
+
+    private function extractDocBlock(\ReflectionProperty $reflectionProperty)
```
Comment (alanpoulain) [inline src/Metadata/Extractor/ClassExtractor.php]: Missing typehint.

File: src/Metadata/Extractor/ClassExtractor.php [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        if (false === $reflectionProperty->getDocComment()) {
+            return [];
+        }
+        $pattern = "#(@[a-zA-Z]+\s*[a-zA-Z0-9, ()_].*)#";
+        preg_match_all($pattern, $reflectionProperty->getDocComment(), $matches, PREG_PATTERN_ORDER);
+
+        return $matches;
+    }
+
+    private function getVarDoc(\ReflectionProperty $reflectionProperty)
```
Comment (alanpoulain) [inline src/Metadata/Extractor/ClassExtractor.php]:
```suggestion
    private function getVarDoc(\ReflectionProperty $reflectionProperty): ?string
```

**Comment by teohhanhui** [comment]
> you don't need the any annotations to expose a simple API, just pure PHP classes.

Shouldn't it use Symfony PropertyInfo then?

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -177,8 +177,9 @@ public function getConfigTreeBuilder()
...
+                        ->booleanNode('scan_dir')->defaultFalse()->end()
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: I don't think we should introduce a new config option. Instead, we should reuse `api_platform.mapping.paths`. Having so many similar options is confusing. We can chain this new extractor with the lowest priority. If necessary, we could provide a toggle to enable/disable this feature.

Comment (flug) [reply src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: You just want to detect the file? I agree with the option that it's potentially too much. After that if you want I can detect if there is the presence of @ApiResource and automatically disable this option. No ?

**Comment by flug** [comment]
@teohhanhui  would you prefer that I use symfony PropertyInfo?

File: src/Metadata/Extractor/ClassExtractor.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Metadata\Extractor;
+
+use ApiPlatform\Core\Util\ReflectionClassRecursiveIterator;
+use Doctrine\Common\Annotations\Reader;
+use Doctrine\ORM\Mapping as ORM;
+use Psr\Container\ContainerInterface;
+
+final class ClassExtractor extends AbstractExtractor
```
Comment (soyuka) [inline src/Metadata/Extractor/ClassExtractor.php]: We shouldn't need this, discussed internally.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php (hunk 9 above)
Comment (teohhanhui) [reply src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: It's possible to use different extractors together, e.g. using annotations together with YAML / XML. That should not be a problem. We do not need to detect the `@ApiResource` annotation. What's important is to make sure that we **do not override** any metadata provided by the user.

File: src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use Symfony\Component\PropertyInfo\PropertyInfoExtractorInterface;
+
+final class DirectoryResourceMetadataFactory implements ResourceMetadataFactoryInterface
+{
+    private $propertyInfo;
+    private $defaults;
+
+    public function __construct(array $defaults = [], PropertyInfoExtractorInterface $propertyInfo)
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php]: should follow the decorator pattern

File: src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    ): ResourceMetadata {
+        $properties = array_map(function (\ReflectionProperty $reflectionProperty) {
+            return $reflectionProperty->getName();
+        }, $reflectionClass->getProperties());
+        $attributes = (null === $properties && [] === $this->defaults['attributes']) ? null :
+            (array)
+            $properties + $this->defaults['attributes'];
+
+        return new ResourceMetadata(
+            $reflectionClass->getShortName(),
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php]: Isn't that the responsibility of https://github.com/api-platform/core/blob/2f941fc3c0dc9dd257da3b34a6bb3f5c21c4230c/src/Metadata/Resource/Factory/ShortNameResourceMetadataFactory.php#L23 ?

File: src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        return $this->createMetadata(new \ReflectionClass($resourceClass));
+    }
+
+    private function createMetadata(
+        \ReflectionClass $reflectionClass
+    ): ResourceMetadata {
+        $properties = array_map(function (\ReflectionProperty $reflectionProperty) {
+            return $reflectionProperty->getName();
+        }, $reflectionClass->getProperties());
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php]: properties should be done in a PropertyNameCollectionMetadataFactory, I think it's already present with the PropertyInfo

File: tests/Metadata/Resource/Factory/DirectoryResourceFilterMetadataFactoryTest.php [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace ApiPlatform\Core\Tests\Metadata\Resource\Factory;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\DirectoryResourceNameCollectionFactory;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceNameCollectionFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceNameCollection;
+use ApiPlatform\Core\Util\ReflectionClassRecursiveIterator;
+use PHPUnit\Framework\TestCase;
+
+class DirectoryResourceFilterMetadataFactoryTest extends TestCase
```
Comment (soyuka) [inline tests/Metadata/Resource/Factory/DirectoryResourceFilterMetadataFactoryTest.php]: DirectoryResourceNameCollectionFactoryTest

File: src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    public function create(string $resourceClass): ResourceMetadata
+    {
+        $parentResourceMetadata = null;
+        if (null !== $this->decorated) {
+            try {
+                $parentResourceMetadata = $this->decorated->create($resourceClass);
+            } catch (ResourceClassNotFoundException $e) {
+            }
+        }
+        $resource['shortName'] = $resource['shortName'] ?? $this->defaults['shortName'] ?? null;
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php]: I don't get where this array is instantiated, could be simplified maybe:
```
$resource = [] + $this->defaults;
```

File: src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+        return $this->update($parentResourceMetadata ?: new ResourceMetadata(), $resource);
+    }
+
+    /**
+     * Creates a new instance of metadata if the property is not already set.
+     */
+    private function update(ResourceMetadata $resourceMetadata, array $metadata): ResourceMetadata
+    {
+        foreach (['shortName'] as $property) {
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php]: do we really need this if there's one property? Let's simplify the code.

File: src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+        $parentResourceMetadata = null;
+        if ($this->decorated) {
+            try {
+                $parentResourceMetadata = $this->decorated->create($resourceClass);
+            } catch (ResourceClassNotFoundException $e) {
+            }
+        }
+        $resource['shortName'] = (new \ReflectionClass($resourceClass))->getShortName();
+
+        return $this->update($parentResourceMetadata ?: new ResourceMetadata(), $resource);
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/DirectoryResourceMetadataFactory.php]:
```
$resourceMetadata = $parentResourceMetadata ?: new ResourceMetadata();

if (null !== $resourceMetadata->getShortname()) {
    return $resourceMetadata;
}

return $resourceMetadata->withShortname((new \ReflectionClass($resourceClass))->getShortName());
```

`update` method isn't required, as isn't the array

File: src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml [hunk 18]
```diff
@@ -14,6 +14,7 @@
         <service id="ApiPlatform\Core\Metadata\Resource\Factory\ResourceNameCollectionFactoryInterface" alias="api_platform.metadata.resource.name_collection_factory" />
 
         <!-- Resource metadata -->
+
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml]: empty line can be removed

**Pull request closed without merging**